<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>An introduction to the cpop package • CPOP</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="An introduction to the cpop package">
<meta property="og:description" content="CPOP">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">CPOP</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.0.24</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/CPOP.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="CPOP_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>An introduction to the cpop package</h1>
                        <h4 data-toc-skip class="author">Kevin Y.X. Wang</h4>
            <address class="author_afil">
      School of Mathematics and Statistics, The University of Sydney, Australia<br><div class="hidden name"><code>CPOP.Rmd</code></div>

    </address>
</div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">CPOP</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tibble.tidyverse.org/" class="external-link">tibble</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## </span>
<span class="co">## Attaching package: 'dplyr'</span></code></pre>
<pre><code><span class="co">## The following objects are masked from 'package:stats':</span>
<span class="co">## </span>
<span class="co">##     filter, lag</span></code></pre>
<pre><code><span class="co">## The following objects are masked from 'package:base':</span>
<span class="co">## </span>
<span class="co">##     intersect, setdiff, setequal, union</span></code></pre>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme_get.html" class="external-link">theme_set</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="fl">18</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>The CPOP package provides functions to perform Cross-Platform Omics Prediction method described in <a href="https://www.biorxiv.org/content/10.1101/2020.12.09.415927v1.full" class="external-link">our paper</a>. The aim is to construct a transferable risk prediction mode with a strong emphasis on addressing practical challenges in the implementation of biomarkers/assays.</p>
<p>Below is a schematic drawing of the CPOP procedure:</p>
<p><img src="1a.png"></p>
<ol style="list-style-type: decimal">
<li><p>Identify a collection of datasets with similar clinical outcomes that can be used for joint modelling.</p></li>
<li><p>Following that, CPOP creates features that are the ratio of each gene’s expression to that of other genes.</p></li>
<li><p>Features that are predictive of the clinical outcome are then selected using a regularised regression modelling framework. These features forms a candidate set of features for further selection.</p></li>
<li><p>The candidate features in Step 3 are reduced to only those features with consistent effect sizes across multiple datasets.</p></li>
<li><p>The final selected features serve as the markers used for final model construction. Visualisation methods are provided.</p></li>
</ol>
</div>
<div class="section level2">
<h2 id="step-1-and-2-data-selection-and-pre-processing">Step 1 and 2: data selection and pre-processing<a class="anchor" aria-label="anchor" href="#step-1-and-2-data-selection-and-pre-processing"></a>
</h2>
<p>The CPOP package provides an example data, <code>cpop_data_binary</code>, to demonstrate how the various functions can be run.</p>
<p><code>cpop_data_binary</code> contains a list of 7 objects:</p>
<ul>
<li>
<code>x1</code>, <code>x2</code>, and <code>x3</code> which are simulated gene expression data.</li>
<li>
<code>y1</code>, <code>y2</code>, and <code>y3</code> which are simulated binary response corresponding to the x’s.</li>
<li>
<code>beta</code> which is the true coefficients of interest.</li>
</ul>
<p>After attaching these objects, we can perform the (log-)ratio constructions using the <code>pairwise_col_diff</code> on the matrices <code>x1</code>, <code>x2</code>, and <code>x3</code>. One of the main idea behind CPOP is that, the gene expression data scale (<code>x1</code>, <code>x2</code>, and <code>x3</code>) can vary greatly between data platforms. But the relative expression between the genes (e.g. difference between gene A and gene B) are more stable. <code>pairwise_col_diff</code> is a function that computes for all pairwise differences between the columns in a given matrix, with the new output being a matrix with the columns <code>geneA--geneB</code>.</p>
<p>Note: the phrase “log-ratio” is used in context where the expression values of biomarkers exhibit differences across several orders of magnitudes. This is the most common in the context of gene expression data with microarray (intensity measurements) and RNA-Sequencing (counts) data. <code>pairwise_col_diff</code> assumes that these values with large magnitude of difference has been dealt with, e.g. through the use of logarithm transformations, and thus, the act of taking difference between columns of a data matrix is similar to how one would construct a ratio (i.e. log(A/B) = log(A) - log(B)). Thus, even though we refer to these values as log-ratios, the actual data matrices that we are operating on are assuming that the columns are already on a transformed scale, such that the difference between feature values is meaningful.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">cpop_data_binary</span>, package <span class="op">=</span> <span class="st">'CPOP'</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">cpop_data_binary</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] "x1"   "x2"   "x3"   "y1"   "y2"   "y3"   "beta"</span></code></pre>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">x1</span> <span class="op">=</span> <span class="va">cpop_data_binary</span><span class="op">$</span><span class="va">x1</span>
<span class="va">x2</span> <span class="op">=</span> <span class="va">cpop_data_binary</span><span class="op">$</span><span class="va">x2</span>
<span class="va">x3</span> <span class="op">=</span> <span class="va">cpop_data_binary</span><span class="op">$</span><span class="va">x3</span>
<span class="va">y1</span> <span class="op">=</span> <span class="va">cpop_data_binary</span><span class="op">$</span><span class="va">y1</span>
<span class="va">y2</span> <span class="op">=</span> <span class="va">cpop_data_binary</span><span class="op">$</span><span class="va">y2</span>
<span class="va">y3</span> <span class="op">=</span> <span class="va">cpop_data_binary</span><span class="op">$</span><span class="va">y3</span>

<span class="va">z1</span> <span class="op">=</span> <span class="fu"><a href="../reference/pairwise_col_diff.html">pairwise_col_diff</a></span><span class="op">(</span><span class="va">x1</span><span class="op">)</span>
<span class="va">z2</span> <span class="op">=</span> <span class="fu"><a href="../reference/pairwise_col_diff.html">pairwise_col_diff</a></span><span class="op">(</span><span class="va">x2</span><span class="op">)</span>
<span class="va">z3</span> <span class="op">=</span> <span class="fu"><a href="../reference/pairwise_col_diff.html">pairwise_col_diff</a></span><span class="op">(</span><span class="va">x3</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/graphics/boxplot.html" class="external-link">boxplot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">x1</span>,<span class="va">x2</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="CPOP_files/figure-html/unnamed-chunk-2-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="step-3-and-4-feature-selection">Step 3 and 4: feature selection<a class="anchor" aria-label="anchor" href="#step-3-and-4-feature-selection"></a>
</h2>
<p>The <code>cpop_model</code> function is the main function of the CPOP package. The first four inputs are <code>z1</code>, <code>z2</code>, <code>y1</code> and <code>y2</code>, where <code>(z1, y1)</code> and <code>(z2, y2)</code> are the log-ratio matrix and the response variable of the first and second data, respectively. In this particular example, the response variable is binary, hence, we can also specify <code>family = "binomial"</code> (default) for this model.</p>
<p>The <code>alpha</code> parameter is one of the penalty terms of the classic Elastic Net model, where <code>alpha = 1</code> corresponds to a Lasso model and <code>alpha = 0</code> corresponds to a Ridge model. The <code>cpop_model</code> uses <code><a href="https://glmnet.stanford.edu/reference/cv.glmnet.html" class="external-link">glmnet::cv.glmnet</a></code> in the backend, and so the penalty term <code>lambda</code> is automatically chosen using the default of <code>lambda.min</code>.</p>
<p>The CPOP model iteratively selects features based on the fit of the model or, if it reaches a pre-determined number of features with the paramter <code>n_features</code> being the main control for this. Here, we arbitary set this to a small number of 10 for illustration.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">cpop_result</span> <span class="op">=</span> <span class="fu"><a href="../reference/cpop_model.html">cpop_model</a></span><span class="op">(</span>z1 <span class="op">=</span> <span class="va">z1</span>, z2 <span class="op">=</span> <span class="va">z2</span>, 
                         y1 <span class="op">=</span> <span class="va">y1</span>, y2 <span class="op">=</span> <span class="va">y2</span>, 
                         family <span class="op">=</span> <span class="st">"binomial"</span>,
                         alpha <span class="op">=</span> <span class="fl">1</span>, 
                         n_features <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## Absolute colMeans difference will be used as the weights for CPOP</span></code></pre>
<pre><code><span class="co">## Fitting CPOP model using alpha = 1</span></code></pre>
<pre><code><span class="co">## Based on previous alpha, 0 features are kept</span></code></pre>
<pre><code><span class="co">## CPOP1 - Step 01: Number of selected features: 0 out of 190</span></code></pre>
<pre><code><span class="co">## CPOP1 - Step 02: Number of selected features: 9 out of 190</span></code></pre>
<pre><code><span class="co">## CPOP1 - Step 03: Number of selected features: 16 out of 190</span></code></pre>
<pre><code><span class="co">## 10 features was reached.</span></code></pre>
<pre><code><span class="co">## A total of 16 features were selected.</span></code></pre>
<pre><code><span class="co">## Removing sources of collinearity gives 13 features.</span></code></pre>
<pre><code><span class="co">## 10 features was reached.</span></code></pre>
<pre><code><span class="co">## A total of 13 features were selected.</span></code></pre>
<pre><code><span class="co">## CPOP2 - Sign: Step 01: Number of leftover features: 9 out of 13</span></code></pre>
<pre><code><span class="co">## The sign matrix between the two data:</span></code></pre>
<pre><code><span class="co">##     </span>
<span class="co">##      -1 0 1</span>
<span class="co">##   -1  0 0 1</span>
<span class="co">##   0   0 0 0</span>
<span class="co">##   1   3 0 0</span></code></pre>
<pre><code><span class="co">## CPOP2 - Sign: Step 02: Number of leftover features: 8 out of 13</span>
<span class="co">## The sign matrix between the two data:</span></code></pre>
<pre><code><span class="co">##     </span>
<span class="co">##      -1 0 1</span>
<span class="co">##   -1  0 0 0</span>
<span class="co">##   0   0 0 0</span>
<span class="co">##   1   1 0 0</span></code></pre>
<pre><code><span class="co">## CPOP2 - Sign: Step 03: Number of leftover features: 8 out of 13</span>
<span class="co">## The sign matrix between the two data:</span></code></pre>
<pre><code><span class="co">##     </span>
<span class="co">##      -1 0 1</span>
<span class="co">##   -1  0 0 0</span>
<span class="co">##   0   0 0 0</span>
<span class="co">##   1   0 0 0</span></code></pre>
</div>
<div class="section level2">
<h2 id="step-5-visualisation">Step 5: visualisation<a class="anchor" aria-label="anchor" href="#step-5-visualisation"></a>
</h2>
<p>The output of the <code>cpop_model</code> function is a <code>cpop</code> object with its own defined methods.</p>
<p>The main method for a <code>cpop</code> object is the <code>plot_cpop</code> function, which produces several useful plot to understand what the CPOP methodology has done or performed.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">cpop_result</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## [1] "cpop" "list"</span></code></pre>
<p>The first plot is using the <code>type = "text"</code> option. A labelled scatter plot shows the estimated coefficients of the selected features from data1 (i.e. <code>(z1, y1)</code>, x-axis) vs the same coefficients from data2 (i.e. <code>(z2, y2)</code>, y-axis).</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plot_cpop.html">plot_cpop</a></span><span class="op">(</span>cpop_result <span class="op">=</span> <span class="va">cpop_result</span>, type <span class="op">=</span> <span class="st">"text"</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## $plot</span></code></pre>
<p><img src="CPOP_files/figure-html/unnamed-chunk-5-1.png" width="700"></p>
<pre><code><span class="co">## </span>
<span class="co">## $data</span>
<span class="co">## <span style="color: #949494;"># A tibble: 8 × 3</span></span>
<span class="co">##   coef_name  coef1    coef2</span>
<span class="co">##   <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="co">## <span style="color: #BCBCBC;">1</span> X01--X10  -<span style="color: #BB0000;">0.322</span> -<span style="color: #BB0000;">0.245</span>  </span>
<span class="co">## <span style="color: #BCBCBC;">2</span> X09--X17   0.722  0.486  </span>
<span class="co">## <span style="color: #BCBCBC;">3</span> X11--X14   0.130  0.001<span style="text-decoration: underline;">58</span></span>
<span class="co">## <span style="color: #BCBCBC;">4</span> X12--X20   0.404  0.152  </span>
<span class="co">## <span style="color: #BCBCBC;">5</span> X01--X07  -<span style="color: #BB0000;">0.437</span> -<span style="color: #BB0000;">0.400</span>  </span>
<span class="co">## <span style="color: #BCBCBC;">6</span> X01--X15  -<span style="color: #BB0000;">0.158</span> -<span style="color: #BB0000;">0.322</span>  </span>
<span class="co">## <span style="color: #BCBCBC;">7</span> X01--X17  -<span style="color: #BB0000;">0.901</span> -<span style="color: #BB0000;">0.599</span>  </span>
<span class="co">## <span style="color: #BCBCBC;">8</span> X04--X12   0.353  0.406</span></code></pre>
<p>We can see that most of the selected features had similar coefficients across the two data. This is exactly what CPOP is aiming to achieve: selecting features that are consistent within each datasets.</p>
<p>However, each of the feature (e.g. <code>X01--X10</code>) actually corresponds to two of the original features in the raw data <code>X01</code> and <code>X10</code> in the data <code>x1</code> and <code>x2</code>. We can further visualise these features in the original features.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plot_cpop.html">plot_cpop</a></span><span class="op">(</span>cpop_result <span class="op">=</span> <span class="va">cpop_result</span>, type <span class="op">=</span> <span class="st">"ggraph"</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## $plot</span></code></pre>
<p><img src="CPOP_files/figure-html/unnamed-chunk-6-1.png" width="700"></p>
<pre><code><span class="co">## </span>
<span class="co">## $data</span>
<span class="co">## <span style="color: #949494;"># A tibble: 8 × 7</span></span>
<span class="co">##   from  to     coef1    coef2 coef_avg coef_abs sign_coef1</span>
<span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     </span>
<span class="co">## <span style="color: #BCBCBC;">1</span> X01   X10   -<span style="color: #BB0000;">0.322</span> -<span style="color: #BB0000;">0.245</span>    -<span style="color: #BB0000;">0.284</span>    0.284  Negative  </span>
<span class="co">## <span style="color: #BCBCBC;">2</span> X09   X17    0.722  0.486     0.604    0.604  Positive  </span>
<span class="co">## <span style="color: #BCBCBC;">3</span> X11   X14    0.130  0.001<span style="text-decoration: underline;">58</span>   0.066<span style="text-decoration: underline;">0</span>   0.066<span style="text-decoration: underline;">0</span> Positive  </span>
<span class="co">## <span style="color: #BCBCBC;">4</span> X12   X20    0.404  0.152     0.278    0.278  Positive  </span>
<span class="co">## <span style="color: #BCBCBC;">5</span> X01   X07   -<span style="color: #BB0000;">0.437</span> -<span style="color: #BB0000;">0.400</span>    -<span style="color: #BB0000;">0.419</span>    0.419  Negative  </span>
<span class="co">## <span style="color: #BCBCBC;">6</span> X01   X15   -<span style="color: #BB0000;">0.158</span> -<span style="color: #BB0000;">0.322</span>    -<span style="color: #BB0000;">0.240</span>    0.240  Negative  </span>
<span class="co">## <span style="color: #BCBCBC;">7</span> X01   X17   -<span style="color: #BB0000;">0.901</span> -<span style="color: #BB0000;">0.599</span>    -<span style="color: #BB0000;">0.750</span>    0.750  Negative  </span>
<span class="co">## <span style="color: #BCBCBC;">8</span> X04   X12    0.353  0.406     0.379    0.379  Positive</span></code></pre>
<p>Here we see that despite the features that we have selected in terms of the ratio, in actual fact, only a handful of features are useful. In particular, <code>X01</code>, <code>X09</code> and <code>X17</code> are the main contributors.</p>
</div>
<div class="section level2">
<h2 id="predictions">Predictions<a class="anchor" aria-label="anchor" href="#predictions"></a>
</h2>
<p>Another useful method on the <code>cpop</code> object is the <code>predict_cpop</code> function. This function takes the previous CPOP result and a new (log-)ratio matrix to produce predictions.</p>
<p>The first column of the output are the sample IDs. The second and third column are prediction results from data1 and data2 respectively. Since we have done the previous diagnostic test that the coefficient size between the two data are very similar, we do expect that these values to be very highly correlated. The last column is <code>cpop_model_avg</code>, which is a simple average between the two predicted values.</p>
<p>Note that the predicted values here are on the logit scale (recall that we are performing a logistic regression on binary responses using the elastic net model in the background). We can also transform these values back to predicted probability scale using the <code><a href="../reference/logit-expit.html">CPOP::expit</a></code> function.</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/predict_cpop.html">predict_cpop</a></span><span class="op">(</span><span class="va">cpop_result</span>, newz <span class="op">=</span> <span class="va">z3</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## <span style="color: #949494;"># A tibble: 6 × 6</span></span>
<span class="co">##   samples cpop_model1 cpop_model2 cpop_model_avg cpop_model_avg_prob</span>
<span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>               <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span>
<span class="co">## <span style="color: #BCBCBC;">1</span> 1             0.465       0.130          0.297               0.573</span>
<span class="co">## <span style="color: #BCBCBC;">2</span> 2            -<span style="color: #BB0000;">1.79</span>       -<span style="color: #BB0000;">0.992</span>         -<span style="color: #BB0000;">1.39</span>                0.207</span>
<span class="co">## <span style="color: #BCBCBC;">3</span> 3             0.853       0.575          0.714               0.671</span>
<span class="co">## <span style="color: #BCBCBC;">4</span> 4            -<span style="color: #BB0000;">0.936</span>      -<span style="color: #BB0000;">0.709</span>         -<span style="color: #BB0000;">0.823</span>               0.306</span>
<span class="co">## <span style="color: #BCBCBC;">5</span> 5            -<span style="color: #BB0000;">0.424</span>      -<span style="color: #BB0000;">0.640</span>         -<span style="color: #BB0000;">0.532</span>               0.370</span>
<span class="co">## <span style="color: #BCBCBC;">6</span> 6             1.00        1.31           1.16                0.760</span>
<span class="co">## <span style="color: #949494;"># … with 1 more variable: cpop_model_avg_class &lt;chr&gt;</span></span></code></pre>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">predicted_plotdf</span> <span class="op">=</span> <span class="fu"><a href="../reference/predict_cpop.html">predict_cpop</a></span><span class="op">(</span><span class="va">cpop_result</span>, newz <span class="op">=</span> <span class="va">z3</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>
    cpop_model1 <span class="op">=</span> <span class="fu"><a href="../reference/logit-expit.html">expit</a></span><span class="op">(</span><span class="va">cpop_model1</span><span class="op">)</span>, 
    cpop_model2 <span class="op">=</span> <span class="fu"><a href="../reference/logit-expit.html">expit</a></span><span class="op">(</span><span class="va">cpop_model2</span><span class="op">)</span>, 
    cpop_model_avg <span class="op">=</span> <span class="fu"><a href="../reference/logit-expit.html">expit</a></span><span class="op">(</span><span class="va">cpop_model_avg</span><span class="op">)</span><span class="op">)</span>

<span class="va">predicted_plotdf</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">cpop_model1</span>, y <span class="op">=</span> <span class="va">cpop_model2</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_abline</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p><img src="CPOP_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
<!-- # Naive Lasso -->
<!-- ```{r} -->
<!-- lasso_result = naive_glmnet(z1, z2, y1, y2, alpha = 1, intercept = FALSE) -->
<!-- lasso_result -->
<!-- ``` -->
<!-- # Plotting -->
<!-- ```{r} -->
<!-- plot_cpop(cpop_result) -->
<!-- plot_cpop(lasso_result) -->
<!-- plot_cpop(cpop_result, type = "ggraph") -->
<!-- plot_cpop(lasso_result, type = "ggraph") -->
<!-- beta -->
<!-- ``` -->
<!-- # Prediction  -->
<!-- ```{r} -->
<!-- # z3 = pairwise_col_diff(x3) -->
<!-- # boxplot(y3 ~ predict_cpop(cpop_result, newz = z3)$cpop_model_avg) -->
<!-- # x3 = matrix(rnorm(n * p, mean = 3, sd = 1), nrow = n, ncol = p) -->
<!-- # colnames(x3) = colnames(x1) -->
<!-- # z3 = pairwise_col_diff(x3) -->
<!-- predict_cpop(cpop_result, newz = z3) %>%  -->
<!--   ggplot(aes(x = cpop_model1, y = cpop_model2)) + -->
<!--   geom_point() + -->
<!--   geom_abline()  -->
<!-- predict_cpop(cpop_result, newz = z3)[,2:3] %>% cor -->
<!-- lasso_pred1 = tibble( -->
<!--   lasso_pred1 = predict(lasso_result$glmnet1, newx = z3), -->
<!--   lasso_pred2 = predict(lasso_result$glmnet2, newx = z3)) -->
<!-- lasso_pred1 %>% cor -->
<!-- lasso_pred1 %>%  -->
<!--   ggplot(aes(x = lasso_pred1, y = lasso_pred2)) + -->
<!--   geom_point() + -->
<!--   geom_abline() -->
<!-- ``` -->
<!-- ```{r} -->
<!-- sessionInfo() -->
<!-- ``` -->
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Kevin Wang.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.9000.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
